### ISSUE TITLE
Редактор скрипт-объектов. Некорректный вывод результатов поиска

### ISSUE THREAD

**Issue by gsbelarus** [GoldenSoftwareLtd/gedemin#4230]
Задал поиск -- строки содержащие латинскую "a". В результатах какие-то лишние пробелы:

![image](https://user-images.githubusercontent.com/5175764/147490634-7d617839-7a0f-4b08-b4f2-0abed1afc134.png)

**Comment by dmitranoid** [comment]
не получилось воспроизвести, 
проблема в одном скрипте ? если да, откуда он взят, попробовать его скопировать- вставить через блокнот, поможет ли

**Comment by gsbelarus** [comment]
Ок. Я перепроверю.

аў, 28 сне 2021, 21:42 карыстальнік dmitranoid ***@***.***>
напісаў:

**Comment by NikolayUkleyko** [comment]
Посмотрел.
Есть такое.

![image](https://user-images.githubusercontent.com/11497453/147640434-b4c3b453-6e94-45c4-8e46-fd46365dc66d.png)
**Рис. 1. Поиск по английской букве "a".**

![image](https://user-images.githubusercontent.com/11497453/147640473-08457b1d-17cf-4ee5-b789-d9f680941b00.png)
**Рис. 2. Результат поиска.**

**Comment by dmitranoid** [comment]
@NikolayUkleyko смущает что только в одной строке, чем она отличается от других, попробуйте скопировать - вставить обратно эту строку через блокнот и повторить поиск

**Comment by NikolayUkleyko** [comment]
@dmitranoid , выполнил копирование буквы "a" (английская) и новый поиск.
Результат, как на рисунке 2, но дополнительно следующее:

![image](https://user-images.githubusercontent.com/11497453/147655319-afea8f28-f1c6-4a0a-a269-c864f1b2cf8a.png)
**Рис. 3. Результат поиска.**

**Comment by dmitranoid** [comment]
@NikolayUkleyko не буквы a, вот эту строку целиком в блокнот и обратно в скрипт чтобы привести ее к ascii, сохранить, повторить поиск
![изображение](https://user-images.githubusercontent.com/3066236/147682284-cf97bf84-6f8f-4a83-b4f6-7617e2d3f355.png)

**Comment by gsbelarus** [comment]
Николай имеет ввиду, что поиск был задан просто по одной букве a. К
найденому нет претензий, есть вопросы к отображению. Обратите внимание,
зачем-то перед а появились пробелы лишние. Вместо Values -- V   alues.

**Comment by dmitranoid** [comment]
да, и это только в одной строке, как и у в вашем примере, у меня было похожее в одном из скриптов был перенос строки CR вместо CR LF и обе строки отображаются в результатах
возможно и эти строки скопированы из источника например в unicode, я не вижу как они хранятся в бд
скопировав скрипт в блокнот и обратно будем уверены что что там только ascii символы, и если это поможет будет понятно что лечить

**Comment by gsbelarus** [comment]
точно ascii. у нас БД в однобайтовой кодировке Win1251 и весь Гедымин. Это
же Делфи 5, там еще не было вменяемого
юникода. единственное место, где в Гедымин есть поддержка юникода это всего
две функции, доступные из
скриптов. одна кодирует из Win1251 в UTF-8 -- другая обратно. они нам нужны
для считывания и записи в JSON.

**Comment by gsbelarus** [comment]
у меня большое подозрение что там был поиск по регулярным выражениям,
например [a], а при выводе на экран
оно пытается выделить кусочек текста для отрисовки жирным шрифтом и берет
длину регулярного выражения,
как буд-то это была подстрока для поиска. Если так, то при включенном флаге
регулярных выражений можно
просто не подсвечивать жирным. так как определить именно подстроку, которая
удовлетворила заданное регулярное
выражение будет сложно.

**Comment by dmitranoid** [comment]
но у @NikolayUkleyko поиск рез regexp (рис.1)
уже думал пересмотреть этот кускок, чтобы показывать всю строку со вхождениями одной строкой, а не как сейчас, каждое вхождение отдельно

**Comment by NikolayUkleyko** [comment]
Обычно поиск выполняется по нескольким символам, фрагменту слова и т.д.
Какой смысл поиска по одному символу, если он может встречаться во многих объектах ?

**Comment by NikolayUkleyko** [comment]
Посмотрел на актуальном **gedemin.exe** — правок для данной темы нет.
Наверное это неактуально. :man_shrugging:

**Comment by NikolayUkleyko** [comment]
@Polfath разбирает данную тему.

**Comment by Polfath** [comment]
После долгого анализа и консультации с Михаилом Юрьевичем предположили, что причина может  крыться в работе самой Delphi5. Предлагаю отложить эту проблему и рассмотреть ее после перехода на Delphi7. Возможно она сама собой исчезнет.

**Comment by TechAnton** [comment]
Посмотрел.

Проблема в переносе строки. Если поиск вернул строку с переносом (2+ строки), то при отрисовке на перенос строки выделяется место, так как это символы, но при выводе на экран они обрезаются и получается пустое пространство, которое вставляется прямо перед найденной строкой. 

<img width="1174" height="716" alt="Image" src="https://github.com/user-attachments/assets/07f0593b-3170-48e8-b211-c08569bb205d" />

## Рубрика "Вопрос-Ответ"
— Почему поиск возвращает 2+ строки? 
— Не знаю ¯\\\_(ツ)\_/¯. На это никто не жаловался, поэтому может так и задумано.

## Технические предложения

Также заметил, что отрисовка не всегда корректно работает и длинные символы не всегда помещаются, так как их длина динамическая. Я попробую это исправить, но если не получится, то есть идея перехода на моноширинный шрифт (и лучше перейти, так как исправления будут через костыли и команды отрисовки). 

Вот несколько примеров поиска по букве "`t`".

#### JetBrains Mono (мой фаворит)
<img width="1695" height="759" alt="Image" src="https://github.com/user-attachments/assets/d503a53b-95b0-4cda-9372-e00868a0c557" />

#### Consolas
<img width="1617" height="758" alt="Image" src="https://github.com/user-attachments/assets/4f5b4e82-feb2-4f32-a849-a07aacd0b07d" />

#### Lucida Console
<img width="1769" height="762" alt="Image" src="https://github.com/user-attachments/assets/b17c9ef7-c368-4cf0-9334-673e2961cbe2" />

#### Source Code Pro
<img width="1721" height="755" alt="Image" src="https://github.com/user-attachments/assets/b1feecf7-c71c-488d-8619-4c07dcd871bb" />

#### Fira Code
<img width="1711" height="753" alt="Image" src="https://github.com/user-attachments/assets/aa0a1d24-2096-45db-b8a2-1c5e213b94e1" />

**Comment by NikolayUkleyko** [comment]
**JetBrains Mono** сбалансированный вариант.

**Comment by TechAnton** [comment]
Посоветовались с коллегами. Решили пока что не менять шрифт, а поменять механизм выделения записи. 
Ранее совпадения выделялись жирным текстом, сейчас это будет зеленая заливка заднего фона.

<img width="575" height="129" alt="Image" src="https://github.com/user-attachments/assets/2373204b-c72d-4cb1-8960-9bc94e0b2f83" />

<img width="1499" height="298" alt="Image" src="https://github.com/user-attachments/assets/1fde9e5a-c50e-4007-bfd5-d7974e47426a" />

**Comment by TechAnton** [comment]
@NikolayUkleyko Готово. 

Ждем обратную связь от пользователей!

**Comment by NikolayUkleyko** [comment]
Проверено.
В целом работает верно.
@TechAnton, пожалуйста, предусмотри возможность отображения уникальных результатов.
То есть, если, искомый фрагмент встречается в нескольких строках объекта, то выводить не каждою строчку, а одну строку с найденным объектом.
В отображение найденного объекта можно добавить ”...”.

**Comment by TechAnton** [comment]
@NikolayUkleyko Выглядит как отдельная огромная тема. 
Ваше предложение перенесено в https://github.com/GoldenSoftwareLtd/gedemin/issues/4768.

**Comment by NikolayUkleyko** [comment]
Проверено.
Работает корректно.
@TechAnton, спасибо !
